idemp-base/src/org/bandahealth/idempiere/base/process/UpdateExpiredStockProcess.java

```diff
@@ -0,0 +1,96 @@
+package org.bandahealth.idempiere.base.process;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.logging.Level;
+
+import org.compiere.model.MAttributeSetInstance;
+import org.compiere.model.MDocType;
+import org.compiere.model.MStorageOnHand;
+import org.compiere.model.Query;
+import org.compiere.process.ProcessInfoParameter;
+import org.compiere.process.StorageCleanup;
+import org.compiere.process.SvrProcess;
+
+/**
+ * This process resets expired stocks to zero for a given client
+ * 
+ * @author clinton
+ *
+ */
+
+public class UpdateExpiredStockProcess extends SvrProcess {
+
+    private final String PROCESS_NAME = this.getClass().getName();
+    public static final String PARAMETERNAME_C_DocType_ID = "C_DocType_ID";
+    
+    @Override
+    protected void prepare() {
+    }
+
+    @Override
+    protected String doIt() throws Exception {
+        log.log(Level.INFO, "START " + PROCESS_NAME);
+
+        int count = 0;
+        //Get all Attribute Set Instance that are expired for this client
+        String attributeWhereClause = MAttributeSetInstance.COLUMNNAME_GuaranteeDate  + " < now()";
+        List<MAttributeSetInstance> expiredAttributes = new Query(getCtx(), MAttributeSetInstance.Table_Name, attributeWhereClause, get_TrxName())
+                .setClient_ID()
+                .setOnlyActiveRecords(true)
+                .list();
+        for (MAttributeSetInstance expiredAttribute : expiredAttributes) {
+            //Get all stocks that have this expired attribute
+            String whereClause = MStorageOnHand.COLUMNNAME_M_AttributeSetInstance_ID + "=?";
+            List<MStorageOnHand> stocks = new Query(getCtx(), MStorageOnHand.Table_Name, whereClause, get_TrxName())
```

Comment:
This will run quite slowly since this query is inside a for loop. Can you just get a list of expired stocks and iterate through them? Something like this:
```
		List<MStorageOnHand> expiredStock = new Query(Env.getCtx(), MStorageOnHand.Table_Name,
				MAttributeSetInstance_BH.Table_Name + "." + MAttributeSetInstance_BH.COLUMNNAME_GuaranteeDate +
						" < now()", get_TrxName()).addJoinClause(
				"JOIN " + MAttributeSetInstance_BH.Table_Name + " ON " + MAttributeSetInstance_BH.Table_Name + "." +
						MAttributeSetInstance_BH.COLUMNNAME_M_AttributeSetInstance_ID + "=" + MStorageOnHand.Table_Name + "." +
						MStorageOnHand.COLUMNNAME_M_AttributeSetInstance_ID).setClient_ID().setOnlyActiveRecords(true).list();
```

Reply:
Actually, we probably need to exclude stock that is `0`.

Reply:
Done

Reply:
Exclude stock that is 0 or equal or less than 0? Thanks